src/Tokenizer/Analyzer/Analysis/AttributeDeclarationAnalysis.php

```diff
@@ -0,0 +1,54 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of PHP CS Fixer.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *     Dariusz Rumiński <dariusz.ruminski@gmail.com>
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace PhpCsFixer\Tokenizer\Analyzer\Analysis;
+
+/**
+ * @internal
+ */
+final class AttributeDeclarationAnalysis
+{
+    private int $startIndex;
+    private int $endIndex;
+    private int $declarationStartIndex;
+    private int $declarationEndIndex;
+
+    public function __construct(int $startIndex, int $endIndex, int $declarationStartIndex, int $declarationEndIndex)
```

Comment:
Let me summarise all my doubts about attribute analysis here, instead of polluting PR with several comments.

I imagine attribute analysis should not require knowing index of opening `#[`. Personally I would rather want to be able to collect attributes for class, property etc., so it would be better to use something like `AttributeAnalyzer::collectFor(int $index)` where `$index` must point to non-whitespace, non-attribute item, and then logic should look for all the attributes _before_ specified index. Probably it's not even required to verify if attributes are placed before supported element, but only rewind up to previous non-whitespace and non-attribute token. I believe this approach would be more reusable in terms of other fixers that would need to know the attributes for elements they process. The reason I'm not fond of current approach is that the current result of attribute analysis depends heavily on how it's invoked - this is very loose contract, which requires internal knowledge outside of the analyzer itself (I mean: fixer right now must navigate through tokens and find _first_ attribute in a group of attributes and use its starting index for analysis, and if you pass starting index of _second_ or consequent attribute, you will get incomplete analysis).

Secondly, I am now sure if `$declarationStartIndex` and `$declarationEndIndex` are required here. Looking at test cases it seems like `$declarationStartIndex` is always the same as `$startIndex`, and `$declarationEndIndex` is always 1 less than `$endIndex`. What's the point of keeping that info? I would rather see it like this:

```php
final class AttributeAnalysis
{
    private int $startIndex;
    private int $endIndex;
    /** @var array<array{start: int, end: int, name: string}> */
    private array $declarations;

    public function __construct(int $startIndex, int $endIndex, array $declarations)
    {
        $this->startIndex = $startIndex;
        $this->endIndex = $endIndex;
        $this->declarations = $declarations;
    }

    public function getStartIndex(): int
    {
        return $this->startIndex;
    }

    public function getEndIndex(): int
    {
        return $this->endIndex;
    }

    public function getDeclarations(): array
    {
        return $this->declarations;
    }
}
```

Mentioned `AttributeAnalyzer::collectFor()` should return `array<AttributeAnalysis>`, and each of the items would represent single `#[]` attribute, with one or more declarations (`$declarations` could also be a collection of some objects, but it's not required with good array shape defined). Concrete example:

```php
<?php

#[Foo]
#[Bar, Baz\Buzz]
function xyz() {}
```

Result of the analysis should be like:

```php
[
    new AttributeAnalysis(2, 4, [
        ['start' => 3, 'end' => 3, 'name' => 'Foo']
    ]),
    new AttributeAnalysis(6, 13, [
        ['start' => 7, 'end' => 7, 'name' => 'Bar'],
        ['start' => 10, 'end' => 12, 'name' => 'Bar\\Buzz']
    ]),
]
```

With this approach attribute analysis could be done only once - right now `applyFix()` calls `AttributeAnalyzer::getAttributeDeclarations()` and then for each `AttributeDeclarationAnalysis` sorting logic calls `AttributeAnalyzer::getAttributes()`, so it seems to me like not optimal solution.

I'll stop review at this point, let me know what you think 🙂.

Reply:
Thanks for the feedback. I haven't gone through it in detail yet, but at first glance, it seems reasonable.

Regarding the `$declarationStartIndex` and `$declarationEndIndex`, `$declarationStartIndex` differs when there are comments before the attribute, while `$declarationEndIndex` includes any whitespace after the attribute and before the subsequent one:
```php
'<?php
/** Start docblock */#[AB\Baz(prop: \'baz\')] #[A\B\Quux(prop1: [1, 2, 4], prop2: true, prop3: \'foo bar\')] #[\A\B\Qux()] #[BarAlias(3)] /** Corge docblock */#[Corge] #[Foo(4, \'baz qux\')]/** End docblock */
function foo() {}
',
3,
[
    new AttributeDeclarationAnalysis(3, 14, 3, 13),
    new AttributeDeclarationAnalysis(15, 48, 15, 47),
    new AttributeDeclarationAnalysis(49, 59, 49, 58),
    new AttributeDeclarationAnalysis(60, 66, 60, 65),
    new AttributeDeclarationAnalysis(67, 71, 68, 70),
    new AttributeDeclarationAnalysis(72, 80, 72, 80),
],
```
I'm not sure if I can get rid of that information.

Reply:
I believe that with the approach suggested above you could get rid of it, because `$declarationStartIndex` would be represented by `start` index of first element in `AttributeAnalysis::$declarations`, similar with `$declarationEndIndex` would be `end` index of the last element. In general, having as an example `#[/* comment */ Foo, Bar(1, 2) /** ... */]`, such analysis would represent something like:
- start index of `#[`
- declarations with already determined name: `/* comment */ Foo` with name `Foo`, `Bar(1, 2) /** ... */` with name `Bar`, both with respective start/end indexes
- end index of `]`

Having such information, you could sort all declarations within `#[]` by pre-parsed `name` contained within each declaration, then replace whole range of tokens inside `#[]`.

```diff
-#[/* comment */ Foo, Bar(1, 2) /** ... */]
+#[Bar(1, 2) /** ... */, /* comment */ Foo]
```

It should be also possible to do something like this:

```diff
-#[Xyz]
-#[/* comment */ Foo, Bar(1, 2) /** ... */]
#[Abc]
+#[Bar(1, 2) /** ... */, /* comment */ Foo]
+#[Xyz]
```

You just would need to compare first declaration of each attribute (after sorting declarations for each attribute) 🙂.

Reply:
Done.

Reply:
Great, I have vacation now, but I'll try to look at it in the meantime 🙂.

Reply:
Continuation [here](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/7395#discussion_r1446300926).